### PR TITLE
fix: Privilege selective conditions over group when building the pre-where clause

### PR DESCRIPTION
**Pull request by fpacifici**
We can often get into a state like this for a snuba query:
```
conditions: [
   ["group_id", "NOT IN", [1,2,3]], # exclude merging groups
   ["issue", "IN", [4]] # filtering from the client
]
```
in the current state the PRE-where may end up including the NOT IN condition and not the IN, which is wrong since the IN is a lot more selective.
This changes the logic that generates the condition based on merged group and ignores it if the user provided something more restrictive.

Unit test added for the corner case.

### CODE REVIEW

File: snuba/query/project_extension.py [hunk 1]
```diff
@@ -112,9 +112,43 @@ def do_post_processing(
...
+                    # Check if there is a more selective conditions around group_id/issue
+                    # in the query conditions. Privilege the positive conditions instead of
+                    # the negative ones.
+                    strict_issue_conditions = False
+                    replaced_conditions: List[Condition] = []
+                    for c in query.get_conditions():
+                        new_condition = c
+                        if (
+                            util.is_condition(c)
+                            and c[0] in ["group_id", "issue"]
```
Comment (fpacifici) [inline snuba/query/project_extension.py]: this is gross. Cannot thing about a better way.

File: snuba/query/project_extension.py [hunk 2]
```diff
@@ -112,9 +112,43 @@ def do_post_processing(
...
-                    query.add_conditions(
-                        [(["assumeNotNull", ["group_id"]], "NOT IN", exclude_group_ids)]
-                    )
+                    # Check if there is a more selective conditions around group_id/issue
+                    # in the query conditions. Privilege the positive conditions instead of
+                    # the negative ones.
+                    strict_issue_conditions = False
+                    replaced_conditions: List[Condition] = []
+                    for c in query.get_conditions():
+                        new_condition = c
```
Comment (tkaemming) [inline snuba/query/project_extension.py]: I find this confusing. Why is this variable needed? Why not just overwrite `c` instead of introducing a new local variable that does the same thing?

File: snuba/query/project_extension.py [hunk 3]
```diff
@@ -112,9 +112,43 @@ def do_post_processing(
...
+                        if (
+                            util.is_condition(c)
```
Comment (tkaemming) [inline snuba/query/project_extension.py]: 😩

File: snuba/query/project_extension.py (hunk 1 above)
Comment (tkaemming) [reply snuba/query/project_extension.py]: I don't have a better idea either (at least one that doesn't require baking in alias awareness into the schema definition, and genericizing this to deal with `IN`/`NOT IN` clauses for any columns.)

This should probably at least comment the abstraction leak/dataset coupling here so we know it's not ideal/intentional rather than leaving it in a review comment.

File: snuba/query/project_extension.py [hunk 4]
```diff
@@ -112,9 +112,43 @@ def do_post_processing(
                 if len(exclude_group_ids) > max_group_ids_exclude:
                     query.set_final(True)
                 else:
-                    query.add_conditions(
-                        [(["assumeNotNull", ["group_id"]], "NOT IN", exclude_group_ids)]
-                    )
+                    # Check if there is a more selective conditions around group_id/issue
+                    # in the query conditions. Privilege the positive conditions instead of
+                    # the negative ones.
+                    strict_issue_conditions = False
```
Comment (tkaemming) [inline snuba/query/project_extension.py]: What is a strict issue condition?

File: snuba/query/project_extension.py [hunk 5]
```diff
@@ -112,9 +112,43 @@ def do_post_processing(
...
+                    for c in query.get_conditions():
+                        new_condition = c
+                        if (
+                            util.is_condition(c)
+                            and c[0] in ["group_id", "issue"]
+                            and c[1] == "IN"
+                            and isinstance(c[2], list)
+                            and all(isinstance(group, int) for group in c[2])
+                        ):
+                            group_ids_with_exclusions = [
```
Comment (tkaemming) [inline snuba/query/project_extension.py]: This name is confusing. "with_exclusions" makes this read like the excluded items are part of this list.

File: snuba/query/project_extension.py [hunk 6]
```diff
@@ -112,9 +112,43 @@ def do_post_processing(
...
+                                group
+                                for group in c[2]
+                                if group not in exclude_group_ids
+                            ]
+                            if group_ids_with_exclusions:
```
Comment (tkaemming) [inline snuba/query/project_extension.py]: If this is an `IN` query that contains 0 items since they were all in the exclusion set, isn't that still significant here?

**Review by tkaemming — DISMISSED** [review]

**Comment by fpacifici** [comment]
Should be done differently now if we go back to this
